### ISSUE TITLE
Failed all tests : No OpKernel was registered to support Op 'BilateralSliceApply' with these attrs.

### ISSUE THREAD

**Issue by dxue2012** [mgharbi/hdrnet#5]
System : Ubuntu 16.04
gcc : 5.4.0
- I used the extra CFLAG `-D _GLIBCXX_USE_CXX11_ABI=0` as pointed out by others who are using GCC version > 5.0

In addition to the requirements listed in `hdrnet/requirements.txt`, I also installed to make things compile:

Cython==0.26
tensorflow=1.1.0
pytest==3.2.1

I was able to run `make` in the hdrnet directory. 

`ls -al build/ lib/` gives:

```
build/:
total 100
drwxr-xr-x 2 root root  4096 Aug 21 05:12 .
drwxr-xr-x 8 root root  4096 Aug 21 05:43 ..
-rw-r--r-- 1 root root 90864 Aug 21 05:12 bilateral_slice.cu.o

lib/:
total 144
drwxr-xr-x 2 root root   4096 Aug 21 05:12 .
drwxr-xr-x 8 root root   4096 Aug 21 05:43 ..
-rwxr-xr-x 1 root root 135408 Aug 21 05:12 hdrnet_ops.so
```

However, after running `py.test test` I get  `11 failed, 3 passed in 2.90 seconds`, where all the failed tests seem to have the same error message (see traceback below). It seems like the Op 'BilateralSliceApply' was not found, despite the files `bilateral_slice.cu.o` and `hdrnet_ops.so` existing.

Do I need to add those files to something like `PATH` or `LD_LIBRARY_PATH`?

Here's the error traceback from one of the tests:
```
_________________________________________________________________ BilateralSliceApplyTest.test_shape_is_correct __________________________________________________________________

self = <hdrnet.test.ops_test.BilateralSliceApplyTest testMethod=test_shape_is_correct>

    def test_shape_is_correct(self):
      batch_size = 3
      grid_shape = [batch_size, 10, 6, 8, 12]
      guide_shape = [batch_size, 101, 60]
      input_shape = [batch_size, 101, 60, 3]
      grid_data = np.random.rand(*grid_shape).astype(np.float32)
      guide_data = np.random.rand(*guide_shape).astype(np.float32)
      input_data = np.random.rand(*input_shape).astype(np.float32)

      for dev in ['/gpu:0']:
>       output_data = self.run_bilateral_slice_apply(dev, grid_data, guide_data, input_data, has_offset=True)

../test/ops_test.py:355:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../test/ops_test.py:341: in run_bilateral_slice_apply
    output_data = sess.run(output_tensor)
/usr/local/lib/python2.7/dist-packages/tensorflow/python/client/session.py:778: in run
    run_metadata_ptr)
/usr/local/lib/python2.7/dist-packages/tensorflow/python/client/session.py:982: in _run
    feed_dict_string, options, run_metadata)
/usr/local/lib/python2.7/dist-packages/tensorflow/python/client/session.py:1032: in _do_run
    target_list, options, run_metadata)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <tensorflow.python.client.session.Session object at 0x7fdb4c551350>, fn = <function _run_fn at 0x7fdb4c2a2488>
args = (<Swig Object of type 'TF_DeprecatedSession *' at 0x7fdb4c397fc0>, {}, ['BilateralSliceApply:0'], [], None, <tensorflow.python.pywrap_tensorflow_internal.TF_Buffer; proxy of <Swig Object of type 'TF_Buffer *' at 0x7fdb4c3979f0> >)
e = InvalidArgumentError()
message = 'No OpKernel was registered to support Op \'BilateralSliceApply\' with these attrs.  Registered devices: [CPU], Regist...'

	 [[Node: BilateralSliceApply = BilateralSliceApply[has_offset=true, _device="/device:GPU:0"](grid, guide, input)]]'
m = <_sre.SRE_Match object at 0x7fdb4c2a2828>

    def _do_call(self, fn, *args):
      try:
        return fn(*args)
      except errors.OpError as e:
        message = compat.as_text(e.message)
        m = BaseSession._NODEDEF_NAME_RE.search(message)
        node_def = None
        op = None
        if m is not None:
          node_name = m.group(1)
          try:
            op = self._graph.get_operation_by_name(node_name)
            node_def = op.node_def
          except KeyError:
            pass
>       raise type(e)(node_def, op, message)
E       InvalidArgumentError: No OpKernel was registered to support Op 'BilateralSliceApply' with these attrs.  Registered devices: [CPU], Registered kernels:
E         device='GPU'
E
E       	 [[Node: BilateralSliceApply = BilateralSliceApply[has_offset=true, _device="/device:GPU:0"](grid, guide, input)]]
E
E       Caused by op u'BilateralSliceApply', defined at:
E         File "/usr/local/bin/py.test", line 11, in <module>
E           sys.exit(main())
E         File "/usr/local/lib/python2.7/dist-packages/_pytest/config.py", line 58, in main
E           return config.hook.pytest_cmdline_main(config=config)
E         File "/usr/local/lib/python2.7/dist-packages/_pytest/vendored_packages/pluggy.py", line 745, in __call__
E           return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)
E         File "/usr/local/lib/python2.7/dist-packages/_pytest/vendored_packages/pluggy.py", line 339, in _hookexec
E           return self._inner_hookexec(hook, methods, kwargs)
E         File "/usr/local/lib/python2.7/dist-packages/_pytest/vendored_packages/pluggy.py", line 334, in <lambda>
E           _MultiCall(methods, kwargs, hook.spec_opts).execute()
E         File "/usr/local/lib/python2.7/dist-packages/_pytest/vendored_packages/pluggy.py", line 614, in execute
E           res = hook_impl.function(*args)
E         File "/usr/local/lib/python2.7/dist-packages/_pytest/main.py", line 139, in pytest_cmdline_main
E           return wrap_session(config, _main)
E         File "/usr/local/lib/python2.7/dist-packages/_pytest/main.py", line 110, in wrap_session
E           session.exitstatus = doit(config, session) or 0
E         File "/usr/local/lib/python2.7/dist-packages/_pytest/main.py", line 146, in _main
E           config.hook.pytest_runtestloop(session=session)
E         File "/usr/local/lib/python2.7/dist-packages/_pytest/vendored_packages/pluggy.py", line 745, in __call__
E           return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)
E         File "/usr/local/lib/python2.7/dist-packages/_pytest/vendored_packages/pluggy.py", line 339, in _hookexec
E           return self._inner_hookexec(hook, methods, kwargs)
E         File "/usr/local/lib/python2.7/dist-packages/_pytest/vendored_packages/pluggy.py", line 334, in <lambda>
E           _MultiCall(methods, kwargs, hook.spec_opts).execute()
E         File "/usr/local/lib/python2.7/dist-packages/_pytest/vendored_packages/pluggy.py", line 614, in execute
E           res = hook_impl.function(*args)
E         File "/usr/local/lib/python2.7/dist-packages/_pytest/main.py", line 169, in pytest_runtestloop
E           item.config.hook.pytest_runtest_protocol(item=item, nextitem=nextitem)
E         File "/usr/local/lib/python2.7/dist-packages/_pytest/vendored_packages/pluggy.py", line 745, in __call__
E           return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)
E         File "/usr/local/lib/python2.7/dist-packages/_pytest/vendored_packages/pluggy.py", line 339, in _hookexec
E           return self._inner_hookexec(hook, methods, kwargs)
E         File "/usr/local/lib/python2.7/dist-packages/_pytest/vendored_packages/pluggy.py", line 334, in <lambda>
E           _MultiCall(methods, kwargs, hook.spec_opts).execute()
E         File "/usr/local/lib/python2.7/dist-packages/_pytest/vendored_packages/pluggy.py", line 613, in execute
E           return _wrapped_call(hook_impl.function(*args), self.execute)
E         File "/usr/local/lib/python2.7/dist-packages/_pytest/vendored_packages/pluggy.py", line 248, in _wrapped_call
E           call_outcome = _CallOutcome(func)
E         File "/usr/local/lib/python2.7/dist-packages/_pytest/vendored_packages/pluggy.py", line 265, in __init__
E           self.result = func()
E         File "/usr/local/lib/python2.7/dist-packages/_pytest/vendored_packages/pluggy.py", line 613, in execute
E           return _wrapped_call(hook_impl.function(*args), self.execute)
E         File "/usr/local/lib/python2.7/dist-packages/_pytest/vendored_packages/pluggy.py", line 248, in _wrapped_call
E           call_outcome = _CallOutcome(func)
E         File "/usr/local/lib/python2.7/dist-packages/_pytest/vendored_packages/pluggy.py", line 265, in __init__
E           self.result = func()
E         File "/usr/local/lib/python2.7/dist-packages/_pytest/vendored_packages/pluggy.py", line 614, in execute
E           res = hook_impl.function(*args)
E         File "/usr/local/lib/python2.7/dist-packages/_pytest/runner.py", line 68, in pytest_runtest_protocol
E           runtestprotocol(item, nextitem=nextitem)
E         File "/usr/local/lib/python2.7/dist-packages/_pytest/runner.py", line 82, in runtestprotocol
E           reports.append(call_and_report(item, "call", log))
E         File "/usr/local/lib/python2.7/dist-packages/_pytest/runner.py", line 162, in call_and_report
E           call = call_runtest_hook(item, when, **kwds)
E         File "/usr/local/lib/python2.7/dist-packages/_pytest/runner.py", line 182, in call_runtest_hook
E           return CallInfo(lambda: ihook(item=item, **kwds), when=when)
E         File "/usr/local/lib/python2.7/dist-packages/_pytest/runner.py", line 196, in __init__
E           self.result = func()
E         File "/usr/local/lib/python2.7/dist-packages/_pytest/runner.py", line 182, in <lambda>
E           return CallInfo(lambda: ihook(item=item, **kwds), when=when)
E         File "/usr/local/lib/python2.7/dist-packages/_pytest/vendored_packages/pluggy.py", line 745, in __call__
E           return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)
E         File "/usr/local/lib/python2.7/dist-packages/_pytest/vendored_packages/pluggy.py", line 339, in _hookexec
E           return self._inner_hookexec(hook, methods, kwargs)
E         File "/usr/local/lib/python2.7/dist-packages/_pytest/vendored_packages/pluggy.py", line 334, in <lambda>
E           _MultiCall(methods, kwargs, hook.spec_opts).execute()
E         File "/usr/local/lib/python2.7/dist-packages/_pytest/vendored_packages/pluggy.py", line 613, in execute
E           return _wrapped_call(hook_impl.function(*args), self.execute)
E         File "/usr/local/lib/python2.7/dist-packages/_pytest/vendored_packages/pluggy.py", line 248, in _wrapped_call
E           call_outcome = _CallOutcome(func)
E         File "/usr/local/lib/python2.7/dist-packages/_pytest/vendored_packages/pluggy.py", line 265, in __init__
E           self.result = func()
E         File "/usr/local/lib/python2.7/dist-packages/_pytest/vendored_packages/pluggy.py", line 614, in execute
E           res = hook_impl.function(*args)
E         File "/usr/local/lib/python2.7/dist-packages/_pytest/runner.py", line 112, in pytest_runtest_call
E           item.runtest()
E         File "/usr/local/lib/python2.7/dist-packages/_pytest/unittest.py", line 176, in runtest
E           self._testcase(result=self)
E         File "/usr/lib/python2.7/unittest/case.py", line 393, in __call__
E           return self.run(*args, **kwds)
E         File "/usr/lib/python2.7/unittest/case.py", line 329, in run
E           testMethod()
E         File "/root/workspace/hdrnet/hdrnet/test/ops_test.py", line 355, in test_shape_is_correct
E           output_data = self.run_bilateral_slice_apply(dev, grid_data, guide_data, input_data, has_offset=True)
E         File "/root/workspace/hdrnet/hdrnet/test/ops_test.py", line 338, in run_bilateral_slice_apply
E           output_tensor = ops.bilateral_slice_apply(grid_tensor, guide_tensor, input_tensor, has_offset=has_offset)
E         File "<string>", line 82, in bilateral_slice_apply
E         File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/framework/op_def_library.py", line 768, in apply_op
E           op_def=op_def)
E         File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/framework/ops.py", line 2336, in create_op
E           original_op=self._default_original_op, op_def=op_def)
E         File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/framework/ops.py", line 1228, in __init__
E           self._traceback = _extract_stack()
E
E       InvalidArgumentError (see above for traceback): No OpKernel was registered to support Op 'BilateralSliceApply' with these attrs.  Registered devices: [CPU], Registered kernels:
E         device='GPU'
E
E       	 [[Node: BilateralSliceApply = BilateralSliceApply[has_offset=true, _device="/device:GPU:0"](grid, guide, input)]]

/usr/local/lib/python2.7/dist-packages/tensorflow/python/client/session.py:1052: InvalidArgumentError
```

**Comment by mgharbi** [comment]
It looks like you are trying to run the tests with no GPU support. 

```
No OpKernel was registered to support Op 'BilateralSliceApply' with these attrs.  Registered devices: [CPU], Registered kernels:
E         device='GPU'
```

We provide only the GPU implementations of the BilateralSlice ops (and related). I might add a CPU version, but that is not high on the priorities currently. Please try to run the tests on a GPU-enabled Tensorflow install.

Hope this helps.

**Comment by dxue2012** [comment]
Thanks for the response! For posterity, here's what I think happened:

1. I did `pip install -r requirements.txt` which only has tensorflow-gpu==1.1.0.
2. `pip` automatically installed tensorflow==1.3.0 since it's a dependency of tensorflow-gpu.
3. `make` complained because tensorflow==1.3.0 is not the version it expected
4. I uninstalled tensorflow 1.30 and installed 1.1.0. This allowed `make` to complete, but because tensorflow-gpu was installed _before_ tensorflow, somehow it wasn't in a good state and thus wasn't available at runtime.
5. (!!) uninstalling both tensorflow and tensorflow-gpu, and re-installing them in the correct order fixes the issue. Yay!
